src/char.cc

```diff
@@ -859,30 +860,35 @@ void en_char_controller::stand(btCollisionWorld *collisionWorld)
     reset(collisionWorld);
 }
 
-
 /* Not part of CC, but reuses the same callbacks etc. */
-phys_ent_ref *
-phys_raycast(glm::vec3 start, glm::vec3 end, btCollisionObject *ignore, btCollisionWorld *world)
+bool
+phys_raycast_entity(glm::vec3 start, glm::vec3 end, btCollisionObject *ignore, btCollisionWorld *world, raycast_info_entity *rc)
 {
+    rc->hit = false;
+
     btVector3 start_bt = glm_to_bt(start);
     btVector3 end_bt = glm_to_bt(end);
     en_ray_result_callback callback(ignore, start_bt, end_bt);
     world->rayTest(start_bt, end_bt, callback);
 
     if (callback.hasHit()) {
-        return (phys_ent_ref *)callback.m_collisionObject->getUserPointer();
+        if (callback.m_collisionObject->getUserPointer()) {
+            rc->hit = true;
+            rc->entity = ((phys_ent_ref *) callback.m_collisionObject->getUserPointer())->ce;
+            rc->hitCoord = glm::vec3(callback.m_hitPointWorld.x(), callback.m_hitPointWorld.y(),
```

Comment:
make this capture the normal as well, and throw away generic

Reply:
(hitting the world is rc->hit = true, rc->entity = invalid)

Reply:
alright. I'll add c_entity::is_valid() as well

Reply:
fixt